### PR TITLE
Externalize forward declarations for Track object

### PR DESCRIPTION
**Pull request by uklotzde**
New file `track_decl.h` with forward declarations and enum definitions for `class Track`.

I picked up this technique from customer projects.

Just a peek of what is actually possible. The majority of #include directives in header files should be replaceable, which in turn will reveal many other missing dependencies like in this PR.

### CODE REVIEW

**Comment by Holzhaus** [comment]
No unnecessary recompiling of large parts of Mixxx if you change `track.h` anymore? Sound good.

**Comment by uklotzde** [comment]
> No unnecessary recompiling of large parts of Mixxx if you change `track.h` anymore? Sound good.

Exactly. Works for all types that are passed around behind shared or managed pointers. Might even reduce compile times in general, but I don't expect much.

**Comment by daschuer** [comment]
> No unnecessary recompiling of large parts of Mixxx if you change track.h anymore? Sounds good.

How should this work? The compiler processes finally a giant cpp file with all includes and macros extracted. 
If track_decl.h is includes into track.h the content of this file will not change.

File: src/library/externaltrackcollection.h [hunk 1]
```diff
@@ -4,12 +4,12 @@
 #include <QList>
 #include <QString>
 
-#include "preferences/usersettings.h"
 #include "library/relocatedtrack.h"
+#include "preferences/usersettings.h"
+#include "track/track_decl.h"
```
Comment (daschuer) [inline src/library/externaltrackcollection.h:9]: Here this will grow the cpp file after prepossessor = more compile time.

**Comment by uklotzde** [comment]
@daschuer So what's your point? I am not willing to start discussions about if a few redundant lines of code in some .cpp files.

I consider this useful and necessary. Otherwise, I will close this PR immediately and we can continue with the old, lazy style of including track.h and some sprinkled forward declarations here and there.

**Comment by daschuer** [comment]
Ah, now I understand. It is useful everywhere you can replace track.h with track_decl.h

File: src/library/externaltrackcollection.h [hunk 2]
```diff
@@ -4,12 +4,12 @@
...
 
 class Library;
 class LibraryFeature;
-class Track;
```
Comment (daschuer) [inline src/library/externaltrackcollection.h:12]: here we should keep just the forward declaration.

File: src/test/soundproxy_test.cpp [hunk 3]
```diff
@@ -4,7 +4,7 @@
 #include "sources/audiosourcestereoproxy.h"
 #include "sources/soundsourceproxy.h"
 #include "test/mixxxtest.h"
-#include "track/trackmetadata.h"
+#include "track/track.h"
```
Comment (daschuer) [inline src/test/soundproxy_test.cpp:7]: Why was this changed?

File: src/track/beatgrid.h [hunk 4]
```diff
@@ -3,9 +3,9 @@
 
 #include <QMutex>
 
-#include "track/track.h"
-#include "track/beats.h"
 #include "proto/beats.pb.h"
+#include "track/beats.h"
+#include "track/track_decl.h"
```
Comment (daschuer) [inline src/track/beatgrid.h:8]: this can become 
class Track;

File: src/track/beatmap.h [hunk 5]
```diff
@@ -10,9 +10,9 @@
 
 #include <QMutex>
 
-#include "track/track.h"
-#include "track/beats.h"
 #include "proto/beats.pb.h"
+#include "track/beats.h"
+#include "track/track_decl.h"
```
Comment (daschuer) [inline src/track/beatmap.h:15]: this can become just 
class Track;

File: src/track/cue.h [hunk 6]
```diff
@@ -6,13 +6,13 @@
...
 #include "audio/types.h"
 #include "track/cueinfo.h"
+#include "track/track_decl.h"
 #include "track/trackid.h"
 #include "util/color/rgbcolor.h"
 #include "util/memory.h"
 
 class CuePosition;
 class CueDAO;
-class Track;
```
Comment (daschuer) [inline src/track/cue.h:15]: please revert.

File: src/track/keyfactory.h [hunk 7]
```diff
@@ -2,12 +2,12 @@
...
 #include <QHash>
-#include <QVector>
 #include <QString>
+#include <QVector>
 
-#include "track/keys.h"
 #include "proto/keys.pb.h"
-#include "track/track.h"
+#include "track/keys.h"
+#include "track/track_decl.h"
```
Comment (daschuer) [inline src/track/keyfactory.h:10]: this can be removed.

**Review by daschuer — CHANGES_REQUESTED** [review]
OK, good idea, thank you. 
I have found a few places where just class Track is sufficient or the new include can be removed entirely.

**Comment by uklotzde** [comment]
I am not going to add forward declarations manually.

**Comment by daschuer** [comment]
In general I think this PR is a good approach. I have just reviewed where track_decl.h is required because of TrackPointer and where not, that's all.
Please reopen.

**Comment by uklotzde** [comment]
I tried to improve the developer experience and to avoid handcrafting the code. But I am not a C++ advocate anymore and won't participate in those discussions.

**Comment by Holzhaus** [comment]
I don't think we need to hand optimize forward declarations. Let's just use `track_decl.h` - the likelyhood that it changes is small (much smaller than the likelyhood that `track.h` changes.

@uklotzde please reopen.

**Comment by daschuer** [comment]
I am working on a new PR on top of this that uses track_decl.h in all headers that uses TrackPointer()

**Pull request closed without merging**